### PR TITLE
Snaji/add benchmarks

### PR DESCRIPTION
**Pull request by slimeth**

### CODE REVIEW

File: tests/common.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 import subprocess
 from typing import List
 
-# test enviroment bucket
+# test environment bucket
```
Comment (radomsak) [inline tests/common.py]: this change should be merged in a separate PR

File: tests/test_live_cp_upload.py [hunk 2]
```diff
@@ -270,6 +270,8 @@ def test_upload_two_files(self):
             text_other_file = content_other_file.download_as_string()
             self.assertEqual(b'hello', text)
             self.assertEqual(b'hello', text_other_file)
+            content.delete()
```
Comment (radomsak) [inline tests/test_live_cp_upload.py]: this change should be merged in a separate PR

File: setup.py [hunk 3]
```diff
@@ -15,11 +15,6 @@
 with open(os.path.join(here, 'README.rst'), encoding='utf-8') as fid:
     long_description = fid.read().strip()  # pylint: disable=invalid-name
 
-with open(os.path.join(here, 'requirements.txt'), encoding='utf-8') as fid:
```
Comment (radomsak) [inline setup.py]: is this change related to the benchmarks? If not it should be a separate PR

File: benchmark/__init__.py [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+import gsutilwrap
+import prettytable
+import temppathlib
+
+import gswrap
+
+# pylint: disable=missing-docstring
+
+
+def print_benchmark(benchmark: str, time_gswrap: float, time_gsutil: float):
```
Comment (radomsak) [inline benchmark/__init__.py]: please add return type -> None here and everywhere in this PR

File: benchmark/__init__.py [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+        self.benchmark_copy_many_files_on_remote()
+        self.benchmark_copy_many_to_many_on_remote_filtered()
+        self.benchmark_rm_many()
+        self.benchmark_read_many()
+        self.benchmark_write_many()
+        self.benchmark_stat_many()
+        self.benchmark_upload_many_to_many_filtered()
+
+    def benchmark_list_many_files(self):
+        for testcase in [10, 1000, 10**4]:
```
Comment (radomsak) [inline benchmark/__init__.py:175]: if `testcase` is the same everywhere it would make sense to have a member variable defining it:

class Benchmark:
    def __init__(self, bucket: str):
        self.testcases = [10, 1000, 10**4]

   for testcase in self.testcases:
      ....

File: benchmark/__init__.py [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+            with temppathlib.TemporaryDirectory() as tmp_dir:
+                for index in range(testcase):
+                    file = tmp_dir.path / "file{}".format(index)
+                    file.write_text("text")
+                try:
+                    self._gsutil_upload_files(path=tmp_dir.path)
```
Comment (radomsak) [inline benchmark/__init__.py]: Comment A:

I read this whole test at once and I was quite puzzled what happens. I saw a function call for _gsutil_upload_files() which does not show where the files are uploaded, and then in line 68 a subprocess deleting files in some scary-looking "self.url_prefix". You should be consistent here, either hide the setup and cleanup in functions or show them explicitly (either delete _gsutil_upload_files or wrap subprocess.check_call[gsutil -m -q rm -r ...] in a function)

File: benchmark/__init__.py [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+                for index in range(testcase):
+                    file = tmp_dir.path / "file{}".format(index)
+                    file.write_text("text")
+                try:
+                    client = gswrap.Client(bucket_name=self.bucket)
+                    gswrap_start = time.time()
+                    self._gswrap_upload_files(client=client, path=tmp_dir.path)
+                    gswrap_end = time.time()
+                finally:
+                    subprocess.check_call([
```
Comment (radomsak) [inline benchmark/__init__.py]: see comment A (here and everywhere same rule applies)

File: benchmark/__init__.py [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+            "gsutil", "-m", "-q", "cp", "-r", "{}".format(path.as_posix()),
+            "{}".format(self.url_prefix)
+        ])
+
+    def benchmark_upload_many_to_many_filtered(self):
+        for testcase in [10, 100, 1000]:
+            with temppathlib.TemporaryDirectory() as tmp_dir:
+                mod = 0
+                for index in range(testcase):
+                    mod += 1
```
Comment (radomsak) [inline benchmark/__init__.py]: if you want to keep this mod operation you should explain in a comment what it does

File: benchmark/__init__.py [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+                        "{}".format(self.url_prefix)
+                    ])
+
+            print_benchmark(
+                benchmark="Benchmark upload-many-to-many {} files".format(
+                    testcase),
+                time_gswrap=gswrap_end - gswrap_start,
+                time_gsutil=gsutil_end - gsutil_start)
+
+    def _gswrap_upload_many_to_many_files(self, client: gswrap.Client,
```
Comment (radomsak) [inline benchmark/__init__.py]: please don't mix benchmarks with helper functions. Just move them up :)

File: benchmark/__init__.py [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+            "gsutil", "-m", "-q", "cp", "-r", "{}".format(self.url_prefix),
+            "{}".format(path.as_posix())
+        ])
+
+    def benchmark_download_many_to_many_filtered(self):
+        for testcase in [10, 100, 1000]:
+            with temppathlib.TemporaryDirectory() as tmp_dir:
+                mod = 0
+                for index in range(testcase):
+                    mod += 1
```
Comment (radomsak) [inline benchmark/__init__.py]: ditto

File: benchmark/benchmark_results [hunk 11]
```diff
@@ -0,0 +1,329 @@
+$ ./benchmark/main.py bucket
```
Comment (radomsak) [inline benchmark/benchmark_results]: reviewed up to here, more on Monday :)

**Review by radomsak — COMMENTED** [review]
Not finished yet :(

File: benchmark/main.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+import argparse
+import sys
+from typing import Any, List
+
+import benchmark
+
+# pylint: disable=missing-docstring
+
+
+class Args:
```
Comment (radomsak) [inline benchmark/main.py:14]: Please rename to `Params` for consistency with the rest of our codebase

```
class Params:
    """Represent parameters parsed from a command line."""

    def __init__(self, args: Any) -> None:
```

File: benchmark/main.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    """Represent parsed command-line arguments."""
+
+    def __init__(self, args: Any) -> None:
+        """Initialize with arguments parsed with ``argparse``."""
```
Comment (radomsak) [inline benchmark/main.py:18]: no need for this comment

File: benchmark/main.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        self.bucket = str(args.bucket)
+
+
+def parse_args(sys_argv: List[str]) -> Args:
```
Comment (radomsak) [inline benchmark/main.py:23]: please move this piece of code in main()

File: benchmark/main.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    parser.add_argument(
+        "bucket",
+        help="Specify name of a accessible "
+        "google cloud storage bucket")
+    args = parser.parse_args(sys_argv[1:])
+
+    return Args(args=args)
+
+
+def _main(args: Args) -> int:
```
Comment (radomsak) [inline benchmark/main.py:38]: Please use our codebase as a pattern how to write main.py functions, for example src/py/batch/logsry.py

File: precommit.py [hunk 16]
```diff
@@ -30,25 +30,28 @@ def main() -> int:
     if overwrite:
         subprocess.check_call([
             "yapf", "--in-place", "--style=style.yapf", "--recursive", "tests",
-            "gswrap", "setup.py", "precommit.py"
+            "gswrap", "benchmark", "setup.py", "precommit.py"
```
Comment (radomsak) [inline precommit.py:33]: I think there is no need to run benchmarks in precommit, let's ask Marko what he thinks

File: precommit.py [hunk 17]
```diff
@@ -30,25 +30,28 @@ def main() -> int:
...
         ],
                               cwd=repo_root.as_posix())
     else:
         subprocess.check_call([
             "yapf", "--diff", "--style=style.yapf", "--recursive", "tests",
-            "gswrap", "bin", "setup.py", "precommit.py"
```
Comment (radomsak) [inline precommit.py:39]: why bin removed?

**Review by radomsak — COMMENTED** [review]
Hi Selim,

I finished the review. The code looks good! There are just some minor comments about the structure.

Could you please think if decorators could make the tests shorter?

Thank you for your work!

Adam

File: tests/common.py (hunk 1 above)
Comment (slimeth) [reply tests/common.py]: Could I add it to the change README / change documentation branch?

File: setup.py (hunk 3 above)
Comment (slimeth) [reply setup.py]: Could I create a branch with multiple minor changes?

File: precommit.py (hunk 17 above)
Comment (slimeth) [reply precommit.py:39]: That was my mistake from a previous change. There is actually no bin directory in the project.
